### PR TITLE
webpack2 (now 3) and some updates

### PR DESCRIPTION
**Pull request by haraldschilly**
issue: #1868 

work in progress, don't merge it

### CODE REVIEW

**Comment by haraldschilly** [comment]
There is at least one non-trivial issue left. It's the use of the message.coffee specification file, which in turn, does load misc, sha1, and finally fails to load base64-js.

```
    ERROR in ./~/buffer/index.js
    Module not found: Error: Can't resolve 'base64-js' in '/projects/14eed217-2d3c-4975-a381-b69edcb40e0e/smc/src/node_modules/buffer'
     @ ./~/buffer/index.js 11:13-33
     @ ./smc-util/~/sha1/sha1.js
     @ ./smc-util/misc.coffee
     @ ./smc-util/message.coffee
     @ ./~/html-webpack-plugin/lib/loader.js!./webapp-lib/doc/api.pug
```

**Comment by williamstein** [comment]
See also https://github.com/sagemathinc/cocalc/issues/2387

**Comment by johnjeng** [comment]
Works in my project. @haraldschilly can you check this?

**Comment by haraldschilly** [comment]
@johnjeng yes, I've looked at this and tested a lot today (e.g. formulas, 3d plots in sagews, latex editor basics, billing page, static page renderings, ...). First I thought we switch to webpack3, but then I realized it's webpack2. So yes, probably better to do this first and once we know it works we go to version 3. I also saw that coffee-script is a bit outdated and inconsistent, so I also updated this. Changing the name to the non-deprecated "coffeescript" without a dash did not work, though. So yes, in summary, webpack-production and -watch do work fine.

However, there is at least one serious issue. When you're in an ipynb file for the new jupyter2 interface, and then go into the menu to switch to the classical one, you end up with:

```
getProjectActions: INVALID project_id -- undefined


AppRedux.getProjectActions | @ | smc-react.coffee?0527:410
-- | -- | --
  | (anonymous) | @ | smc-react.coffee?0527:27
  | exports.JupyterActions.JupyterActions.file_action | @ | actions.coffee?0474:1355
  | (anonymous) | @ | actions.coffee?0474:1
  | confirm_dialog.cb | @ | actions.coffee?0474:1860
  | store.wait.cb | @ | actions.coffee?0474:1512
  | (anonymous) | @ | smc-react.coffee?0527:185
```

Hmm ... I haven't checked if this also happens in production, though. Just mentioning this here to avoid us getting overly excited about webpack2.

**Comment by williamstein** [comment]
>  I haven't checked if this also happens in production, though.

"File --> Switch to classical notebook" I think works fine in prod right now.

**Comment by johnjeng** [comment]
>Changing the name to the non-deprecated "coffeescript" without a dash did not work,

Yes, that is coffeescript 2. I expect A LOT to break from that. It also includes CJSX support out of the box which is one of my motivating factors to update webpack first.

**Comment by williamstein** [comment]
>  CJSX support

Awesome too, because it is a new supported version, rather than a broken unsupported version...

**Comment by johnjeng** [comment]
Alright, there is just one issue left I am aware of: 

Switching to Classical Jupyter doesn't redraw the editor again properly. Navigating to any tab and coming back makes it redraw correctly.

Switching from Classical Jupyter to Cocalc Jupyter works no problem (as expected due to awesomeness of react...)

File: src/smc-webapp/jupyter/actions.coffee [hunk 1]
```diff
@@ -1346,14 +1346,22 @@ class exports.JupyterActions extends Actions
...
     # The path may optionally be *any* file in this project.
     file_action: (action_name, path) =>
         a = @redux.getProjectActions(@store.get('project_id'))
         path ?= @store.get('path')
+        if action_name == 'reopen_file'
+            a.close_file(path)
+            # ensure the side effects from changing registered
+            # editors in project_file.coffee finish happening
+            window.setTimeout =>
+                a.open_file(path: path)
```
Comment (johnjeng) [inline src/smc-webapp/jupyter/actions.coffee]: I think it's interesting that this seems necessary... If this is the real fix then the current code shouldn't work at all. Maybe it has something to do with how `require` is resolved by upgrading webpack.

**Comment by johnjeng** [comment]
@williamstein I can't figure out where this loading text comes from. It doesn't seem to be the classical Jupyter editor, the new jupyter editor, nor the `ProjectContentViewer`. It's also strange that there's no spinner like in react editors. It's not like non-react editors which are an alert. 

screen shot 2017-12-22 at 4 44 40 pm

**Comment by williamstein** [comment]
> Any ideas?

It's jquery, so it is probably in a .html file, and gets revealed via a .show() or hidden.

**Comment by johnjeng** [comment]
Well, as far as I know this works now... But I would prefer to address this timeout strangeness.

**Comment by johnjeng** [comment]
Seems to work for me on initial inspection. Needs an in-depth test by @haraldschilly

**Comment by haraldschilly** [comment]
Ahh, I understand. There is also something kucalc related in the "site settings" for the administrator.

**Comment by johnjeng** [comment]
To clarify, you're looking into this issue, correct?

**Comment by haraldschilly** [comment]
Yes, I'm still on this.

**Comment by haraldschilly** [comment]
Now node8 (newer npm?) issues pop up. I have impression it's a dorman problem appearing now, not a fundamental problem. Errors start like that:

```
ERROR in ./node_modules/parse-asn1/index.js
Module not found: Error: Can't resolve 'browserify-aes' in '/home/user/smc/src/node_modules/parse-asn1'
 @ ./node_modules/parse-asn1/index.js 4:14-39
 @ ./node_modules/browserify-sign/browser/verify.js
 @ ./node_modules/browserify-sign/browser/index.js
 @ (webpack)/node_modules/crypto-browserify/index.js
 @ ./smc-webapp/widget-ssh-keys/fingerprint.coffee
 @ ./smc-webapp/widget-ssh-keys/main.cjsx
 @ ./smc-webapp/account_ssh_keys.cjsx
 @ ./smc-webapp/account_page.cjsx
 @ ./smc-webapp/mobile_app.cjsx
 @ ./smc-webapp/entry-point.coffee
 @ ./webapp-smc.coffee
 
ERROR in ./node_modules/parse-asn1/fixProc.js
Module not found: Error: Can't resolve 'browserify-aes' in '/home/user/smc/src/node_modules/parse-asn1'
 @ ./node_modules/parse-asn1/fixProc.js 6:14-39
 @ ./node_modules/parse-asn1/index.js
 @ ./node_modules/browserify-sign/browser/verify.js
 @ ./node_modules/browserify-sign/browser/index.js
 @ (webpack)/node_modules/crypto-browserify/index.js
 @ ./smc-webapp/widget-ssh-keys/fingerprint.coffee
 @ ./smc-webapp/widget-ssh-keys/main.cjsx
 @ ./smc-webapp/account_ssh_keys.cjsx
 @ ./smc-webapp/account_page.cjsx
 @ ./smc-webapp/mobile_app.cjsx
 @ ./smc-webapp/entry-point.coffee
 @ ./webapp-smc.coffee

[...]
```

I've added crypto browserify here 4bae8fa8d70cfccd567f27bc544ee2b85c59833f and well, I suspect something changed. The crypto library is directly part of node now, ...

**Pull request merged**